File: 02-abstract-factory/02-abstract-factory.js
class Service {
    createSender() {
        throw new Error("not implemented");
    }
    createLogger() {
        throw new Error("not implemented");
    }
}

class EmailService extends Service {
    createSender() {
        return new EmailSender();
    }

    createLogger() {
        return new EmailLogger();
    }
}

class PushService extends Service {
    createSender() {
        return new PushSender();
    }

    createLogger() {
        return new PushLogger();
    }
}

class EmailSender {
    send() {
        return "sending email"
    }
}

class EmailLogger {
    log() {
        return "email log saved"
    }
}

class PushSender {
    send() {
        return "push notification sent"
    }
}

class PushLogger {
    log() {
        return "push log saved"
    }
}

class Client {
    #sender;
    #logger;

    constructor(service) {
        this.#sender = service.createSender();
        this.#logger = service.createLogger();
    }

    process() {
        console.log(this.#sender.send())
        console.log(this.#logger.log())
    }
}

c = new Client(new EmailService())
c.process()

File: 02-abstract-factory/02-abstract-factory.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Protocol

class Service(ABC):
    @abstractmethod
    def create_sender(self) -> MessageSender:
        pass

    @abstractmethod
    def create_logger(self) -> MessageLogger:
        pass

class EmailService(Service):
    def create_sender(self) -> MessageSender:
        return EmailSender()

    def create_logger(self) -> MessageLogger:
        return EmailLogger()

class SMSService(Service):
    def create_sender(self) -> MessageSender:
        return SMSSender()

    def create_logger(self) -> MessageLogger:
        return SMSLogger()

class MessageSender(Protocol):
    def send(self) -> None:
        pass

class MessageLogger(Protocol):
    def log(self) -> None:
        pass

class EmailSender(MessageSender):
    def send(self) -> None:
        print("email has been sent")

class EmailLogger(MessageLogger):
    def log(self) -> None:
        print("email has been logged")

class SMSSender(MessageSender):
    def send(self) -> None:
        print("sms has been sent")

class SMSLogger(MessageLogger):
    def log(self) -> None:
        print("sms has been logged")

class Client:
    def __init__(self, service: Service) -> None:
        self.sender = service.create_sender()
        self.logger = service.create_logger()

    def process(self) -> None:
        self.sender.send()
        self.logger.log()

def main():
    emailService = EmailService()
    client = Client(emailService)
    client.process()

    smsService = SMSService()
    client = Client(smsService)
    client.process()

if __name__ == "__main__":
    main()

